ocdata/core/bulk.py

```diff
@@ -0,0 +1,287 @@
+import os
+import pickle
+
+import ase
+import numpy as np
+import pymatgen
+from pymatgen.core.surface import (
+    SlabGenerator,
+    get_symmetrically_distinct_miller_indices,
+)
+from pymatgen.io.ase import AseAtomsAdaptor
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
+
+from ocdata.configs.constants import COVALENT_MATERIALS_MPIDS, MAX_MILLER
+from ocdata.configs.paths import BULK_PKL_PATH, PRECOMPUTED_SLABS_DIR_PATH
+
+
+class Bulk:
+    """
+    Initializes a bulk object in one of 3 ways:
+    - Directly pass in an ase.Atoms object.
+    - Pass in index of bulk to select from bulk database.
+    - Randomly sample a bulk from bulk database.
+
+    Arguments
+    ---------
+    bulk_atoms: ase.Atoms
+        Bulk structure.
+    bulk_id_from_db: int
+        Index of bulk to select if not doing a random sample.
+    bulk_db_path: str
+        Path to bulk database.
+    precomputed_slabs_path: str
+        Path to folder of precomputed slabs.
+    mp_id: str
+        Materials Project ID of bulk to select if not doing a random sample.
+        [CURRENTLY NOT SUPPORTED.]
+    """
+
+    def __init__(
+        self,
+        bulk_atoms: ase.Atoms = None,
+        bulk_id_from_db: int = None,
+        # mp_id = None,
+        bulk_db_path: str = BULK_PKL_PATH,
+        precomputed_slabs_path: str = PRECOMPUTED_SLABS_DIR_PATH,
+    ):
+        self.bulk_id_from_db = bulk_id_from_db
+        self.bulk_db_path = bulk_db_path
+        self.precomputed_slabs_path = precomputed_slabs_path
+
+        if bulk_atoms is not None:
+            self.atoms = bulk_atoms.copy()
+            self.mpid = None
+        elif bulk_id_from_db is not None:
+            bulk_db = pickle.load(open(bulk_db_path, "rb"))
+            (
+                self.atoms,
+                self.mpid,
+                _,
+                _,
+            ) = bulk_db[bulk_id_from_db]
```

Comment:
This expects the database to have 4 entries for each index, 2 of which aren't used. DCGAT, and likely future bulks don't necessarily have this structure.

1. Do we just keep this structure and expect someone to create a database identical to this format?
2. Instead specify `bulk_info = bulk_db[bulk_id_from_db]` which is a tuple of arbitrary length (some databases may have more/less information). The only requirement here would then be that `self.atoms = bulk_info[0]` and `self.bulk_id = bulk_info[1]`. We can place asserts here as well to make sure new databases conform with that.


Slight preference towards the latter.

Reply:
Yep, we should redo the database files a bit. I'd much rather switch to a tuple of (ase.Atoms, dict), where dict can have an arbitrary number of attributes, and only a subset of which would be required to initialize bulks in ocdata. That way we're not limiting flexibility and this pipeline still works if the pkl file has more attributes saved.

Reply:
Yup that sounds great. This would break compatibility with the existing databases but I think it's fine to update those as part of this refactor.

Reply:
I updated the bulk pkl in the repo to be a list of dictionaries with each dict containing atoms objects, mpid, and other metadata: https://github.com/Open-Catalyst-Project/Open-Catalyst-Dataset/blob/refactor-v2/ocdata/core/bulk.py#L49